src/zino/tasks/linkstatetask.py

```diff
@@ -0,0 +1,141 @@
+from zino.statemodels import EventState, InterfaceState, Port, PortStateEvent
+from zino.tasks.task import Task
+
+_logger = logging.getLogger(__name__)
+
+BASE_POLL_LIST = ("ifIndex", "ifDescr", "ifAlias", "ifAdminStatus", "ifOperStatus", "ifLastChange")
+
+
+@dataclass
+class BaseInterfaceRow:
+    index: int
+    descr: str
+    alias: str
+    admin_status: str
+    oper_status: str
+    last_change: int
+
+    def is_sane(self) -> bool:
+        return bool(self.index and self.descr)
+
+
+class LinkStateTask(Task):
+    """Fetches and stores state information about router ports/links.
+
+    Things Zino 1 does at this point that this implementation ignores:
+
+    1. Zino 1 would fetch and record OLD-CISCO-INTERFACES-MIB::locIfReason, but this isn't very useful for anything
+    other than very old equipment.
+
+    2. Zino 1 collects and records interface stacking/layering information from IF-MIB::ifStackTable. It was used to
+    deem an interface as either significant or insignificant for making events about.  It was used because Uninett's
+    old convention was to set interface descriptions on only the sub-unit of Juniper ports, but this is no longer the
+    case: Descriptions are mandated for both physical ports and their sub-units.
+    """
+
+    async def run(self):
+        snmp = SNMP(self.device)
+        poll_list = [("IF-MIB", column) for column in BASE_POLL_LIST]
+        attrs = await snmp.sparsewalk(*poll_list)
+        _logger.debug("%s ifattrs: %r", self.device.name, attrs)
+
+        self._update_interfaces(attrs)
+
+    def _update_interfaces(self, new_attrs: SparseWalkResponse):
+        for index, row in new_attrs.items():
+            self._update_single_interface(row)
+
+    def _update_single_interface(self, row: dict[str, Any]):
+        data = BaseInterfaceRow(*(row.get(attr) for attr in BASE_POLL_LIST))
+        if not data.is_sane():
+            return
+
+        port = self._get_or_create_port(data.index)
+        port.ifdescr = data.descr
+        self._update_ifalias(port, data)
+
+        if not self._is_interface_watched(data):
+            return
+
+        self._update_state(data, port, row)
+
+    def _update_state(self, data: BaseInterfaceRow, port: Port, row: dict[str, Any]):
+        for attr in ("ifAdminStatus", "ifOperStatus"):
+            if not row.get(attr):
+                raise MissingInterfaceTableData(self.device.name, data.index, attr)
+
+        state = f"admin{data.admin_status.capitalize()}"
+        # A special tweak so that we report ports in oper-down (but admin-up) state first time we see them
+        if not port.state and data.oper_status != "up" and state != "adminDown":
+            port.state = InterfaceState.UNKNOWN
+        if state == "adminUp":
+            state = data.oper_status
+        state = InterfaceState(state)
+        if port.state and port.state != state:
+            self._make_or_update_state_event(port, state)
+        port.state = state
```

Comment:
I feel this is a bit hard to read, with `port.state` and `state` being used everywhere and set to different values several times. Even though its a pretty short function, I had to look at this pretty hard to understand whats happening

Reply:
How do you think I felt when I read the original Tcl code this was translated from? :-D